design/decisions/0004-entity-dregister-api.md

```diff
@@ -0,0 +1,147 @@
+# Entity Deregister API
+
+* Date: __2024-03-14__
+* Status: __New__
+
+## Background
+
+All the entities (devices and services) registered with %%te%% are stored with the MQTT broker as retained messages,
+with their metadata spread across multiple topics.
+For example, for a child device `child01`, the registration message is stored on the `te/device/child01//` topic,
+its twin data is stored across several `te/device/child01///twin/<twin-data-type>` topics,
+its command metadata stored across several `te/device/child01///cmd/<cmd-type>` topics and so on.
+
+On top of that, when a device has services associated with it, or has nested child devices,
+they all have their own respective registration topics and respective metadata topics.
+So, deregistering a device involves deregistering itself, its metadata and
+the complete entity hierarchy that is associated with it.
+
+## Challenges
+
+* Metadata and data associated with a given entity spread across multiple topics, must be cleared together.
+* When services and child devices are linked to a device, deleting that entity involves deleting all linked entities,
+  to avoid leaving orphan entities behind.
+* The parent and child devices may not share a common topic prefix, making it difficult to do it purely from an MQTT level.
+  For example, a device `te/device/child01//` may have an associated service `te/device/child01/service/service01`
+  and nested child devices `te/device/child11//` and `te/device/child12//`.
+  Even though the parent device and service has a common prefix `te/device/child01/#`, there is no such common prefix
+  between the parent device and its child devices.
+  And when custom topic schemes are used, even the device and its service may not have any common prefix available.
+
+## Solution Proposals
+
+### Using third-party MQTT tools like mosquitto_sub
+
+The `mosquitto_sub` command provides options like `--remove-retained` to selectively remove all retained messages in a topic hierarchy by using topic filters.
+For example, the following command can be used to remove all retained messages under the `te/device/child01` topic tree:
+
+```sh
+mosquitto_sub --remove-retained -t "te/device/child01/+/+/#"
+```
+
+The mappers must also be updated to remove entities from their entity store on receipt of these clear messages.
```

Comment:
Currently, the mapper just ignores empty registration messages. So, this mapper change is a must-have regardless of the solution that we pick.

Reply:
Since the mappers are responsible for the interactions with the clouds, they should also be configurable when removing the device from the cloud as sometimes you only want to delete the local data and want to keep the data in the cloud (e.g. measurements from a decommissioned device), and since not deleting the data from the cloud is less destructive, then that should be the default (to avoid unexpected data loss).